x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/autoscaling/AutoscalerService.java

```diff
@@ -0,0 +1,218 @@
+    private final ClusterService clusterService;
+    private final Client client;
+    private final boolean isNlpEnabled;
+
+    private final Map<String, Stats> lastInferenceStatsByDeploymentNode;
+    private final Map<String, Autoscaler> autoscalers;
+
+    private volatile Scheduler.Cancellable cancellable;
+
+    public AutoscalerService(ThreadPool threadPool, ClusterService clusterService, Client client, boolean isNlpEnabled) {
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.isNlpEnabled = isNlpEnabled;
+
+        lastInferenceStatsByDeploymentNode = new HashMap<>();
+        autoscalers = new HashMap<>();
+    }
+
+    public void start() {
+        update(clusterService.state());
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        update(event.state());
+    }
+
+    private synchronized void update(ClusterState state) {
+        if (isNlpEnabled == false) {
+            return;
+        }
+
+        TrainedModelAssignmentMetadata assignments = TrainedModelAssignmentMetadata.fromState(state);
+        for (TrainedModelAssignment assignment : assignments.allAssignments().values()) {
+            if (assignment.getAutoscalingSettings() != null && assignment.getAutoscalingSettings().getEnabled()) {
+                Autoscaler autoscaler = autoscalers.computeIfAbsent(
+                    assignment.getDeploymentId(),
+                    key -> new Autoscaler(assignment.getDeploymentId(), assignment.totalTargetAllocations())
+                );
+                autoscaler.setMinMaxNumberOfAllocations(
+                    assignment.getAutoscalingSettings().getMinNumberOfAllocations(),
+                    assignment.getAutoscalingSettings().getMaxNumberOfAllocations()
+                );
+            } else {
+                autoscalers.remove(assignment.getDeploymentId());
+            }
+        }
+
+        if (autoscalers.isEmpty() == false) {
+            if (cancellable == null) {
+                logger.debug("Starting ML inference autoscaler");
+                scheduleNextTrigger();
+            }
+        } else {
+            stop();
+        }
+    }
+
+    public synchronized void stop() {
+        if (cancellable != null && cancellable.isCancelled() == false) {
+            logger.debug("Stopping ML inference autoscaler");
+            cancellable.cancel();
+            cancellable = null;
+        }
+    }
+
+    private void scheduleNextTrigger() {
+        try {
+            cancellable = threadPool.schedule(this::trigger, TimeValue.timeValueSeconds(TIME_INTERVAL_SECONDS), threadPool.generic());
+        } catch (EsRejectedExecutionException e) {
+            if (e.isExecutorShutdown() == false) {
+                throw e;
+            }
+        }
+    }
+
+    private synchronized void trigger() {
+        scheduleNextTrigger();
```

Comment:
Best to schedule the next after the processing has completed. In a slow cluster it could take more than 10 second to get and process the deployment stats. You can use `ActionListener.runAfter()` to call `scheduleNextTrigger()`

Reply:
I had refactored this to using `threadPool.scheduleWithFixedDelay` and an `AtomicBoolean busy` to bail out if it's still busy.